libcontainer/rootfs_linux.go

```diff
@@ -1115,3 +1119,53 @@ func setRecAttr(m *configs.Mount, rootfs string) error {
 		return unix.MountSetattr(-1, procfd, unix.AT_RECURSIVE, m.RecAttr)
 	})
 }
+
+// validateRoot detects symlinks in config.Rootfs that may break config.ReadonlyPaths
+// and config.MaskPaths.
+//
+// For preventing CVE-2023-27561 (CVE-2019-19921 re-introduction/regression).
+// https://github.com/opencontainers/runc/issues/3751
+func validateRoot(config *configs.Config) error {
+	mustNotBeSymlink := make(map[string]struct{})
+	for _, f := range append(config.ReadonlyPaths, config.MaskPaths...) {
+		mustNotBeSymlink[f] = struct{}{}
+		a, err := ancestorPaths(f)
+		if err != nil {
+			return fmt.Errorf("failed to compute the ancestor paths of %q: %w", f, err)
+		}
+		for _, g := range a {
+			mustNotBeSymlink[g] = struct{}{}
+		}
+	}
+	// mustNotBeSymlink typically contains {"/sys", "/proc"} in addition to readonlyPaths and maskedPaths.
+	rootFD, err := unix.Open(config.Rootfs, unix.O_DIRECTORY|unix.O_RDONLY, 0o600)
+	if err != nil {
+		return err
+	}
+	defer unix.Close(rootFD)
+	for f := range mustNotBeSymlink {
+		// fstatat(2): "If pathname is absolute, then dirfd is ignored."
+		fRel := strings.TrimPrefix(f, string(os.PathSeparator))
```

Comment:
As the path is absolute and cleaned, it would be safe to use `f[1:]` in the call to `Fstatat`.

Reply:
For readability I don't prefer to use`f[1:]`

Reply:
The alternative is using `"."+f` in call to Fstatat but I don't like it either.